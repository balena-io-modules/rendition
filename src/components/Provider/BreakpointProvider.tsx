import debounce from 'lodash/debounce';
import * as React from 'react';
import { BreakpointContext } from '../../contexts/BreakpointContext';
export interface BreakpointProviderProps extends React.PropsWithChildren<{}> {
	breakpoints: number[];
	onBreakpointChange?: (breakpoint: number) => void;
}

export const BreakpointProvider: React.FC<BreakpointProviderProps> = ({
	breakpoints,
	children,
}) => {
	const [currentBreakpoint, setCurrentBreakpoint] = React.useState<number>(0);

	React.useEffect(() => {
		if (!breakpoints || !breakpoints.length) {
			return;
		}

		const handleResize = () => {
			const clientWidth = window.innerWidth;
			const breakpointIndex = breakpoints.findIndex(
				(breakpoint) => clientWidth < breakpoint,
			);
			const currentBreakpoint =
				breakpointIndex >= 0 ? breakpointIndex : breakpoints.length - 1;
			setCurrentBreakpoint(currentBreakpoint);
		};

		const debouncedResize = debounce(handleResize, 80);

		handleResize();

		window.addEventListener('resize', debouncedResize, { passive: false });

		return () => {
			window.removeEventListener('resize', debouncedResize);
		};
	}, [breakpoints]);

	return (
		<BreakpointContext.Provider value={{ currentBreakpoint }}>
			{children}
		</BreakpointContext.Provider>
	);
};

BreakpointProvider.displayName = 'BreakpointProvider';
