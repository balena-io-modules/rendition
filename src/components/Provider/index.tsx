import { Grommet } from 'grommet';
import merge from 'lodash/merge';
import * as React from 'react';
import styled from 'styled-components';
import { DefaultProps, Theme } from '../../common-types';
import defaultTheme from '../../theme';
import { px } from '../../utils';

const Base = styled(Grommet)`
	font-family: ${props => props.theme.font};
	font-size: ${props => px(props.theme.fontSizes[1])};

	h1,
	h2,
	h3,
	h4,
	h5,
	h6,
	button {
		font-family: ${props => props.theme.titleFont};
	}
`;

const Provider = ({ theme, ...props }: ThemedProvider) => {
	return <Base theme={merge(defaultTheme, theme)} {...props} />;
};

export interface ThemedProvider extends DefaultProps {
	theme?: Theme;
}

export default Provider;
