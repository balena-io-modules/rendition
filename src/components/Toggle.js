import * as React from 'react'
import styled, { withTheme } from 'styled-components'
import { lighten, getColor } from '../utils'
import theme from '../theme'
import { compose } from 'recompose'
import asRendition from '../asRendition'
/**
 *  -  toggle styles copied and modified from https://codepen.io/mallendeo/pen/eLIiG
 *     by Mauricio Allende (https://mallendeo.com/)
 *
 *  -  Toggle switch html structure inspired by Kent C. Odds - Advanced React Patterns V2
 *     https://codesandbox.io/s/github/kentcdodds/advanced-react-patterns-v2/tree/egghead/?from-embed
 *
 * @todo add the source url in Toggle.md
 */
const Button = styled.button`
  box-sizing: initial;
  display: inline-block;
  outline: 0;
  width: ${({ emphazised }) => (emphazised ? '4em' : '2em')};
  height: ${({ emphazised }) => (emphazised ? '2em' : '1em')};
  position: relative;
  cursor: pointer;
  user-select: none;
  background: #fbfbfb;
  border-radius: ${({ emphazised }) => (emphazised ? '4em' : '2em')};
  padding: 4px;
  transition: all 0.4s ease;
  border: ${({ emphazised }) => (emphazised ? '2px' : '1px')} solid #e8eae9;

  &:focus::after,
  &:active::after {
    box-sizing: initial;
  }

  &::after {
    left: 0;
    position: relative;
    display: block;
    content: '';
    width: 50%;
    height: 100%;
    border-radius: ${({ emphazised }) => (emphazised ? '4em' : '2em')};
    background: #fbfbfb;
    transition: all 0.3s cubic-bezier(0.175, 0.885, 0.32, 1.275),
      padding 0.3s ease, margin 0.3s ease;
    box-shadow: 0 0 0 2px rgba(0, 0, 0, 0.1), 0 4px 0 rgba(0, 0, 0, 0.08);
  }
  &:active::after {
    padding-right: 1.6em;
  }
  ${props =>
    props.on &&
    `
    &::after {
      left: 50%;
    }
    & {
      background: ${props.activeBg ||
        getColor(props, 'bg', 'main') ||
        '#86d993'};
    }
    &:disabled {
      background: ${lighten(
    props.activeBg || (getColor(props, 'bg', 'main') || '#86d993')
  )};
    }
    &:active {
      box-shadow: none;
    }
    &:active::after {
      margin-left: -1.6em;
    }
  `};
`
Button.defaultProps = { theme }

const Input = styled.input`
  display: none;
`

class Toggle extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      on: props.on
    }
    this.handleClick = this.handleClick.bind(this)
  }

  handleClick () {
    this.setState(
      ({ on }) => {
        return { on: !on }
      },
      () => {
        this.props.onToggle(this.state.on)
      }
    )
  }

  render () {
    const { disabled, emphazised, ...props } = this.props
    return (
      <span>
        <Input type='checkbox' checked={this.state.on} onChange={() => {}} />
        <Button
          {...props}
          aria-label='Toggle'
          disabled={disabled}
          onClick={this.handleClick}
          on={this.state.on}
          emphazised={emphazised}
        />
      </span>
    )
  }
}

export default compose(withTheme, asRendition)(Toggle)
