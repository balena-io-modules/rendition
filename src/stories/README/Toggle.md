# Toggle

Displays a toggle switch. 

If neither the `activeBg` prop nor a theme is provided, the default 'active' background color `#86d993` is applied. 

## Props

| Name          | Type      | Default   | Required   | Description                                          |
| ------ | ------ | --------- | ---------- | ------------- |
| `activeBg`| `string`  | -         | -          | The background color of the toggle when switched on. It takes the precedence over the theme color                     |
| `primary`     | `boolean` | -         | -          | If true, use the `primary` theme color               |
| `secondary`   | `boolean` | -         | -          | If true, use the `secondary` theme color             |
| `tertiary`    | `boolean` | -         | -          | If true, use the `tertiary` theme color              |
| `quarternary` | `boolean` | -         | -          | If true, use the `quarternary` theme color           |
| `danger`      | `boolean` | -         | -          | If true, use the `danger` theme color                |
| `warning`     | `boolean` | -         | -          | If true, use the `warning` theme color               |
| `success`     | `boolean` | -         | -          | If true, use the `success` theme color               |
| `info`        | `boolean` | -         | -          | If true, use the `info` theme color                  |
| `emphasized`  | `boolean` | -         | -          | If true, use a larger size                           |
| `on`  | `boolean`  | -         | -          | The state of the toggle switch        |
| `onToggle`    | `() => void` | -         | -          | Callback invoked when the "toggle" is switched, it gets passed the swith's state |