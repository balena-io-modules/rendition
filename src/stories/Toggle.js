import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import withReadme from 'storybook-readme/with-readme'
import { Box, Txt, Toggle, Provider } from '../'
import * as Readme from './README/Toggle.md'

storiesOf('Core/Toggle', module)
  .addDecorator(withReadme(Readme))
  .add('Standard', () => {
    return (
      <Provider>
        <Box m={3}>
          <Toggle mx={2} on onToggle={action('case-1')} />
          <Txt.span>No theme</Txt.span>
        </Box>
        <Box m={3}>
          <Toggle mx={2} on primary onToggle={action('case-2')} />
          <Txt.span>Primary</Txt.span>
        </Box>
        <Box m={3}>
          <Toggle mx={2} on secondary onToggle={action('case-3')} />
          <Txt.span>Secondary</Txt.span>
        </Box>
        <Box m={3}>
          <Toggle mx={2} on tertiary onToggle={action('case-4')} />
          <Txt.span>Tertiary</Txt.span>
        </Box>
        <Box m={3}>
          <Toggle mx={2} on quartenary onToggle={action('case-5')} />
          <Txt.span>Quartenary</Txt.span>
        </Box>
        <Box m={3}>
          <Toggle mx={2} on success onToggle={action('case-6')} />
          <Txt.span>Success</Txt.span>
        </Box>
        <Box m={3}>
          <Toggle mx={2} on danger onToggle={action('case-7')} />
          <Txt.span>Danger</Txt.span>
        </Box>
        <Box m={3}>
          <Toggle mx={2} on warning onToggle={action('case-8')} />
          <Txt.span>Warning</Txt.span>
        </Box>
        <Box m={3}>
          <Toggle mx={2} on info onToggle={action('case-9')} />
          <Txt.span>Info</Txt.span>
        </Box>
      </Provider>
    )
  })
  .add('Emphazised', () => {
    return (
      <Provider>
        <Box m={3}>
          <Toggle on={false} emphazised onToggle={action('case-1')} />
        </Box>
      </Provider>
    )
  })
  .add('Custom toggled backgound', () => {
    return (
      <Provider>
        <Box m={3}>
          <Toggle
            mr={2}
            on
            activeBg={'palevioletred'}
            onToggle={action('case-1')}
          />
          <Txt.span>Enabled</Txt.span>
        </Box>
        <Box m={3}>
          <Toggle
            mr={2}
            on
            activeBg={'palevioletred'}
            disabled
            onToggle={action('case-2')}
          />
          <Txt.span>Disabled</Txt.span>
        </Box>
      </Provider>
    )
  })
  .add('Disabled', () => {
    return (
      <Provider>
        <Box m={3}>
          <Toggle mr={2} on onToggle={action('case-1')} />
          <Txt.span>Enabled</Txt.span>
        </Box>
        <Box m={3}>
          <Toggle mr={2} on disabled onToggle={action('case-2')} />
          <Txt.span>Disabled</Txt.span>
        </Box>
      </Provider>
    )
  })
