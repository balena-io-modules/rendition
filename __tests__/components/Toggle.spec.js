/* globals expect, test */
import React from 'react'
import renderer from 'react-test-renderer'
import Provider from '../../src/components/Provider'
import Toggle from '../../src/components/Toggle'

test('ToggleButton renders correctly', () => {
  const component = renderer.create(
    <Provider>
      <Toggle />
    </Provider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
