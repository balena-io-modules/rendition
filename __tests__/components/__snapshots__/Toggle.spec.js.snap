// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleButton renders correctly 1`] = `
<div
  className="sc-bdVaJa jKopLr"
>
  <span>
    <input
      className="sc-htoDjs lflUtN"
      onChange={[Function]}
      type="checkbox"
    />
    <button
      aria-label="Toggle"
      className="sc-dnqmqq gDpXJY sc-gzVnrw kxwHSE"
      onClick={[Function]}
    />
  </span>
</div>
`;
